File: src/apps/EventViewer/src/EventQueue.m
#import "EventQueue.h"
#import "KarabinerKit/KarabinerKit.h"
#import "PreferencesKeys.h"
#import "libkrbn/libkrbn.h"
#import <pqrs/weakify.h>

@interface EventQueue ()

@property NSMutableArray* queue;
@property NSMutableDictionary<NSNumber*, NSMutableSet<NSString*>*>* modifierFlags;
@property(copy) NSString* simpleModificationJsonString;
@property(weak) IBOutlet NSTableView* view;
@property(weak) IBOutlet NSButton* addSimpleModificationButton;

- (void)pushKeyEvent:(NSString*)code
                name:(NSString*)name
           eventType:(NSString*)eventType
                misc:(NSString*)misc;
- (void)updateAddSimpleModificationButton:(NSString*)title;

@end

static void hid_value_observer_callback(uint64_t device_id,
                                        enum libkrbn_hid_value_type type,
                                        uint32_t value,
                                        enum libkrbn_hid_value_event_type event_type,
                                        void* refcon) {
  EventQueue* queue = (__bridge EventQueue*)(refcon);
  if (!queue) {
    return;
  }

  @weakify(queue);
  dispatch_async(dispatch_get_main_queue(), ^{
    @strongify(queue);
    if (!queue) {
      return;
    }

    NSNumber* deviceId = @(device_id);

    NSString* code = nil;
    if ([[NSUserDefaults standardUserDefaults] boolForKey:kShowHex]) {
      code = [NSString stringWithFormat:@"0x%02x", value];
    } else {
      code = [NSString stringWithFormat:@"%d", value];
    }
    NSString* keyType = @"";
    NSMutableDictionary* simpleModificationJson = [NSMutableDictionary new];

    char buffer[256];
    buffer[0] = '\0';
    NSString* name = @"";

    switch (type) {
      case libkrbn_hid_value_type_key_code: {
        keyType = @"key";
        libkrbn_get_key_code_name(buffer, sizeof(buffer), value);
        name = [NSString stringWithUTF8String:buffer];

        NSNumber* unnamedNumber = nil;
        {
          uint32_t unnamedKeyCodeNumber = 0;
          if (libkrbn_find_unnamed_key_code_number(buffer, &unnamedKeyCodeNumber)) {
            unnamedNumber = [NSNumber numberWithInteger:unnamedKeyCodeNumber];
          }
        }

        if (libkrbn_is_modifier_flag(value)) {
          NSMutableSet* set = queue.modifierFlags[deviceId];
          if (!set) {
            set = [NSMutableSet new];
            queue.modifierFlags[deviceId] = set;
          }

          if (event_type == libkrbn_hid_value_event_type_key_down) {
            [set addObject:name];
          } else {
            [set removeObject:name];
          }
        }

        simpleModificationJson[@"from"] = [NSMutableDictionary new];
        simpleModificationJson[@"from"][@"key_code"] = unnamedNumber ? unnamedNumber : name;
        break;
      }

      case libkrbn_hid_value_type_consumer_key_code: {
        keyType = @"consumer_key";
        libkrbn_get_consumer_key_code_name(buffer, sizeof(buffer), value);
        name = [NSString stringWithUTF8String:buffer];

        NSNumber* unnamedNumber = nil;
        {
          uint32_t unnamedConsumerKeyCodeNumber = 0;
          if (libkrbn_find_unnamed_consumer_key_code_number(buffer, &unnamedConsumerKeyCodeNumber)) {
            unnamedNumber = [NSNumber numberWithInteger:unnamedConsumerKeyCodeNumber];
          }
        }

        simpleModificationJson[@"from"] = [NSMutableDictionary new];
        simpleModificationJson[@"from"][@"consumer_key_code"] = unnamedNumber ? unnamedNumber : name;
        break;
      }
    }

    NSString* eventType = @"";
    switch (event_type) {
      case libkrbn_hid_value_event_type_key_down:
        eventType = [NSString stringWithFormat:@"%@_down", keyType];
        break;
      case libkrbn_hid_value_event_type_key_up:
        eventType = [NSString stringWithFormat:@"%@_up", keyType];
        break;
      case libkrbn_hid_value_event_type_single:
        eventType = @"";
        break;
    }

    NSString* misc = @"";
    {
      NSMutableSet* set = queue.modifierFlags[deviceId];
      if (set && set.count > 0) {
        NSArray* flags = [[set allObjects] sortedArrayUsingSelector:@selector(compare:)];
        misc = [misc stringByAppendingString:[NSString stringWithFormat:@"flags: %@ ", [flags componentsJoinedByString:@","]]];
      }
    }

    [queue pushKeyEvent:code name:name eventType:eventType misc:misc];

    if (simpleModificationJson.count > 0) {
      queue.simpleModificationJsonString = [KarabinerKitJsonUtility createJsonString:simpleModificationJson];
      [queue updateAddSimpleModificationButton:[NSString stringWithFormat:@"Add `%@` to Karabiner-Elements", name]];
    }
  });
}

@implementation EventQueue

enum {
  MAXNUM = 256,
};

- (instancetype)init {
  self = [super init];

  if (self) {
    _queue = [NSMutableArray new];
    _modifierFlags = [NSMutableDictionary new];
  }

  return self;
}

- (void)dealloc {
  [self terminateHidValueObserver];

  [[NSDistributedNotificationCenter defaultCenter] removeObserver:self
                                                             name:nil
                                                           object:nil];
}

- (void)setup {
  [self initializeHidValueObserver];
  [self updateAddSimpleModificationButton:nil];

  {
    NSString* name = [NSString stringWithUTF8String:libkrbn_get_distributed_notification_device_grabbing_state_is_changed()];
    NSString* object = [NSString stringWithUTF8String:libkrbn_get_distributed_notification_observed_object()];

    [[NSDistributedNotificationCenter defaultCenter] addObserver:self
                                                        selector:@selector(deviceGrabbingStateIsChangedCallback)
                                                            name:name
                                                          object:object
                                              suspensionBehavior:NSNotificationSuspensionBehaviorDeliverImmediately];
  }
}

- (void)initializeHidValueObserver {
  libkrbn_enable_hid_value_monitor(hid_value_observer_callback,
                                   (__bridge void*)(self));
}

- (void)terminateHidValueObserver {
  libkrbn_disable_hid_value_monitor();
}

- (void)updateAddSimpleModificationButton:(NSString*)title {
  if (title) {
    self.addSimpleModificationButton.hidden = NO;
  } else {
    self.addSimpleModificationButton.hidden = YES;
  }
  self.addSimpleModificationButton.title = title;
}

- (void)deviceGrabbingStateIsChangedCallback {
  [self initializeHidValueObserver];
}

- (BOOL)observed {
  return libkrbn_hid_value_monitor_observed();
}

- (NSInteger)numberOfRowsInTableView:(NSTableView*)aTableView {
  return [self.queue count];
}

- (id)tableView:(NSTableView*)aTableView objectValueForTableColumn:(NSTableColumn*)aTableColumn row:(NSInteger)rowIndex {
  id identifier = [aTableColumn identifier];

  NSDictionary* dict = self.queue[([self.queue count] - 1 - rowIndex)];
  return dict[identifier];
}

- (void)refresh {
  [self.view reloadData];
  [self.view scrollRowToVisible:([self.queue count] - 1)];
}

- (NSString*)modifierFlagsToString:(NSUInteger)flags {
  NSMutableArray* names = [NSMutableArray new];
  if (flags & NSEventModifierFlagCapsLock) {
    [names addObject:@"caps"];
  }
  if (flags & NSEventModifierFlagShift) {
    [names addObject:@"shift"];
  }
  if (flags & NSEventModifierFlagControl) {
    [names addObject:@"ctrl"];
  }
  if (flags & NSEventModifierFlagOption) {
    [names addObject:@"opt"];
  }
  if (flags & NSEventModifierFlagCommand) {
    [names addObject:@"cmd"];
  }
  if (flags & NSEventModifierFlagNumericPad) {
    [names addObject:@"numpad"];
  }
  if (flags & NSEventModifierFlagHelp) {
    [names addObject:@"help"];
  }
  if (flags & NSEventModifierFlagFunction) {
    [names addObject:@"fn"];
  }

  return [names componentsJoinedByString:@","];
}

- (NSString*)buttonToString:(NSEvent*)event {
  NSInteger number = [event buttonNumber];
  return [NSString stringWithFormat:@"button%d", (int)(number + 1)];
}

- (void)pushKeyEvent:(NSString*)code
                name:(NSString*)name
           eventType:(NSString*)eventType
                misc:(NSString*)misc {
  [self push:eventType
        code:code
        name:name
        misc:misc];
}

- (void)pushMouseEvent:(NSEvent*)event eventType:(NSString*)eventType {
  NSString* flags = [self modifierFlagsToString:[event modifierFlags]];
  [self push:eventType
        code:[NSString stringWithFormat:@"%d", (int)([event buttonNumber])]
        name:[self buttonToString:event]
        misc:[NSString stringWithFormat:@"{x:%d,y:%d} click_count:%d %@",
                                        (int)([event locationInWindow].x), (int)([event locationInWindow].y),
                                        (int)([event clickCount]),
                                        [flags length] > 0 ? [NSString stringWithFormat:@"flags:%@", flags] : @""]];
}

- (void)pushScrollWheelEvent:(NSEvent*)event eventType:(NSString*)eventType {
  [self push:eventType
        code:@""
        name:@""
        misc:[NSString stringWithFormat:@"dx:%.03f dy:%.03f dz:%.03f", [event deltaX], [event deltaY], [event deltaZ]]];
}

- (void)pushMouseEvent:(NSEvent*)event {
  switch ([event type]) {
    case NSEventTypeLeftMouseDown:
    case NSEventTypeRightMouseDown:
    case NSEventTypeOtherMouseDown:
      [self pushMouseEvent:event eventType:@"button_down"];
      break;

    case NSEventTypeLeftMouseUp:
    case NSEventTypeRightMouseUp:
    case NSEventTypeOtherMouseUp:
      [self pushMouseEvent:event eventType:@"button_up"];
      break;

    case NSEventTypeLeftMouseDragged:
    case NSEventTypeRightMouseDragged:
    case NSEventTypeOtherMouseDragged:
      [self pushMouseEvent:event eventType:@"mouse_dragged"];
      break;

    case NSEventTypeScrollWheel:
      [self pushScrollWheelEvent:event eventType:@"scroll_wheel"];
      break;

    default:
      // Do nothing
      break;
  }
}

- (void)push:(NSString*)eventType code:(NSString*)code name:(NSString*)name misc:(NSString*)misc {
  NSDictionary* dict = @{@"eventType" : eventType,
                         @"code" : code,
                         @"name" : name,
                         @"misc" : misc};

  [self.queue insertObject:dict atIndex:0];
  if ([self.queue count] > MAXNUM) {
    [self.queue removeLastObject];
  }
  [self refresh];
}

- (IBAction)clear:(id)sender {
  [self.queue removeAllObjects];
  [self updateAddSimpleModificationButton:nil];
  [self refresh];
}

- (IBAction)copy:(id)sender {
  NSPasteboard* pboard = [NSPasteboard generalPasteboard];
  NSMutableString* string = [NSMutableString new];

  for (NSUInteger i = 0; i < [self.queue count]; ++i) {
    NSDictionary* dict = self.queue[([self.queue count] - 1 - i)];

    NSString* eventType = [NSString stringWithFormat:@"type:%@", dict[@"eventType"]];
    NSString* code = [NSString stringWithFormat:@"code:%@", dict[@"code"]];
    NSString* name = [NSString stringWithFormat:@"name:%@", dict[@"name"]];
    NSString* misc = [NSString stringWithFormat:@"misc:%@", dict[@"misc"]];

    [string appendFormat:@"%@ %@ %@ %@\n",
                         [eventType stringByPaddingToLength:20
                                                 withString:@" "
                                            startingAtIndex:0],
                         [code stringByPaddingToLength:15
                                            withString:@" "
                                       startingAtIndex:0],
                         [name stringByPaddingToLength:20
                                            withString:@" "
                                       startingAtIndex:0],
                         misc];
  }

  if ([string length] > 0) {
    [pboard clearContents];
    [pboard writeObjects:@[ string ]];
  }
}

- (IBAction)addSimpleModification:(id)sender {
  NSString* escapedJsonString = [self.simpleModificationJsonString stringByAddingPercentEncodingWithAllowedCharacters:NSCharacterSet.URLQueryAllowedCharacterSet];
  NSString* url = [NSString stringWithFormat:@"karabiner://karabiner/simple_modifications/new?json=%@", escapedJsonString];
  NSLog(@"url %@", url);
  [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:url]];
}

@end

File: src/apps/EventViewer/src/EventQueue.swift
import SwiftUI

private func callback(_ deviceId: UInt64,
                      _ usagePage: Int32,
                      _ usage: Int32,
                      _ eventType: libkrbn_hid_value_event_type,
                      _ context: UnsafeMutableRawPointer?)
{
    let obj: EventQueue! = unsafeBitCast(context, to: EventQueue.self)

    DispatchQueue.main.async { [weak obj] in
        guard let obj = obj else { return }

        if !libkrbn_is_momentary_switch_event(usagePage, usage) {
            return
        }

        var buffer = [Int8](repeating: 0, count: 256)
        let entry = EventQueueEntry()

        //
        // entry.code
        //

        if UserSettings.shared.showHex {
            entry.usagePage = String(format: "0x%02x", usagePage)
            entry.usage = String(format: "0x%02x", usage)
        } else {
            entry.usagePage = String(format: "%d", usagePage)
            entry.usage = String(format: "%d", usage)
        }

        //
        // entry.name
        //

        libkrbn_get_momentary_switch_event_json_string(&buffer, buffer.count, usagePage, usage)
        let jsonString = String(cString: buffer)

        entry.name = jsonString

        //
        // modifierFlags
        //

        if libkrbn_is_modifier_flag(usagePage, usage) {
            libkrbn_get_modifier_flag_name(&buffer, buffer.count, usagePage, usage)
            let modifierFlagName = String(cString: buffer)

            if obj.modifierFlags[deviceId] == nil {
                obj.modifierFlags[deviceId] = Set()
            }

            if eventType == libkrbn_hid_value_event_type_key_down {
                obj.modifierFlags[deviceId]!.insert(modifierFlagName)
            } else {
                obj.modifierFlags[deviceId]!.remove(modifierFlagName)
            }
        }

        //
        // entry.eventType
        //

        switch eventType {
        case libkrbn_hid_value_event_type_key_down:
            entry.eventType = "down"
        case libkrbn_hid_value_event_type_key_up:
            entry.eventType = "up"
        case libkrbn_hid_value_event_type_single:
            break
        default:
            break
        }

        //
        // entry.misc
        //

        if let set = obj.modifierFlags[deviceId] {
            if set.count > 0 {
                let flags = set.sorted().joined(separator: ",")
                entry.misc = "flags \(flags)"
            }
        }

        //
        // EventQueue.queue
        //

        obj.append(entry)

        //
        // simpleModificationJsonString
        //

        libkrbn_get_simple_modification_json_string(&buffer, buffer.count, usagePage, usage)
        let simpleModificationJsonString = String(cString: buffer)

        if simpleModificationJsonString != "" {
            obj.simpleModificationJsonString = simpleModificationJsonString

            libkrbn_get_momentary_switch_event_usage_name(&buffer, buffer.count, usagePage, usage)
            let usageName = String(cString: buffer)
            obj.addSimpleModificationButtonText = "Add \(usageName) to Karabiner-Elements"
        }
    }
}

public class EventQueueEntry: Identifiable, Equatable {
    public var id = UUID()
    public var eventType = ""
    public var usagePage = ""
    public var usage = ""
    public var name = ""
    public var misc = ""

    public static func == (lhs: EventQueueEntry, rhs: EventQueueEntry) -> Bool {
        lhs.id == rhs.id
    }
}

public class EventQueue: ObservableObject {
    public static let shared = EventQueue()

    let maxCount = 256
    var modifierFlags: [UInt64: Set<String>] = [:]

    @Published var queue: [EventQueueEntry] = []
    @Published var simpleModificationJsonString: String = ""
    @Published var addSimpleModificationButtonText: String = ""
    @Published var unknownEventEntries: [EventQueueEntry] = []

    init() {
        clear()

        let obj = unsafeBitCast(self, to: UnsafeMutableRawPointer.self)
        libkrbn_enable_hid_value_monitor(callback, obj)
    }

    deinit {
        libkrbn_disable_hid_value_monitor()
    }

    public func observed() -> Bool {
        libkrbn_hid_value_monitor_observed()
    }

    public func append(_ entry: EventQueueEntry) {
        queue.append(entry)
        if queue.count > maxCount {
            queue.removeFirst()
        }
    }

    public func clear() {
        queue.removeAll()

        // Fill queue with empty entries to avoid SwiftUI List rendering corruption at MainView.swift.
        while queue.count < maxCount {
            queue.append(EventQueueEntry())
        }
    }

    public func copyToPasteboard() {
        var string = ""

        queue.forEach { entry in
            if entry.eventType.count > 0 {
                let eventType = "type:\(entry.eventType)".padding(toLength: 20, withPad: " ", startingAt: 0)
                let code = "HID usage: \(entry.usagePage),\(entry.usage)".padding(toLength: 20, withPad: " ", startingAt: 0)
                let name = "name:\(entry.name)".padding(toLength: 60, withPad: " ", startingAt: 0)
                let misc = "misc:\(entry.misc)"

                string.append("\(eventType) \(code) \(name) \(misc)\n")
            }
        }

        if !string.isEmpty {
            let pboard = NSPasteboard.general
            pboard.clearContents()
            pboard.writeObjects([string as NSString])
        }
    }

    public func addSimpleModification() {
        guard let string = simpleModificationJsonString.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) else { return }
        guard let url = URL(string: "karabiner://karabiner/simple_modifications/new?json=\(string)") else { return }
        NSWorkspace.shared.open(url)
    }

    //
    // Mouse event handling
    //

    func modifierFlagsString(_ flags: NSEvent.ModifierFlags) -> String {
        var names: [String] = []
        if flags.contains(.capsLock) {
            names.append("caps")
        }
        if flags.contains(.shift) {
            names.append("shift")
        }
        if flags.contains(.control) {
            names.append("ctrl")
        }
        if flags.contains(.option) {
            names.append("opt")
        }
        if flags.contains(.command) {
            names.append("cmd")
        }
        if flags.contains(.numericPad) {
            names.append("numpad")
        }
        if flags.contains(.help) {
            names.append("help")
        }
        if flags.contains(.function) {
            names.append("fn")
        }

        return names.joined(separator: ",")
    }
}
